### PR TITLE
refactor(32118) Redesign the propagation of the workspace status

### PR DESCRIPTION
**Pull request by vanch3d**
**Kanban Ticket:** https://hivemq.kanbanize.com/ctrl_board/57/cards/32118/details/

---

## Description

This PR transforms how users understand the state of their data transformation pipelines in the HiveMQ Edge Workspace. Previously, users could only see if nodes were connected or disconnected. Now, users can instantly see both **runtime status** (is it connected?) and **operational status** (is it configured?) through intuitive visual feedback.

The enhancement introduces:

- **Color-coded connection states**: Green = running, Red = error, Yellow = stopped
- **Animated edges for configured paths**: Flowing dots indicate data transformations are ready to work
- **Per-connection feedback**: Each data path shows its own configuration state independently
- **Instant problem detection**: Immediately see which paths need configuration attention

### User Experience Improvements

**What users gain:**

- **At-a-glance status understanding**: No need to click into nodes to check configuration
- **Clear visual language**: Color shows runtime state, animation shows configuration completeness
- **Faster troubleshooting**: Immediately identify which connections need attention
- **Reduced configuration errors**: Visual feedback prevents incomplete setups
- **Confidence in deployment**: See that all paths are both connected AND configured before going live

### Technical Summary

**Implementation highlights:**

- Integrated dual-status model across all 10 workspace node types
- Implemented 8 per-edge operational status rules for accurate feedback
- Migrated to React Flow's efficient hooks for better performance
- Added 74+ comprehensive tests (unit, component, and E2E)
- Full documentation in workspace topology reference

---

### Previous Behavior - Limited Status Information

The old implementation only showed basic connection status:

**Limitations:**

- Only runtime status visible (connected vs disconnected)
- No indication if data transformations were configured
- All edges to the same target looked identical
- Users had to manually check each node to verify configuration
- No visual distinction between "connected but not configured" and "fully operational"
- Difficult to spot incomplete configurations at scale

HiveMQ-Edge-10-27-2025_07_34_PM

_Example: All edges appear the same regardless of configuration state_

---

### New Behavior - Rich Visual Feedback

The new implementation provides comprehensive status visualization with two dimensions:

#### 1. Workspace with Mixed Status States

When viewing a complex workspace with various configuration states:

HiveMQ-Edge-10-27-2025_07_33_PM

_Test: Captured from Cypress E2E test `workspace-pr-screenshots.spec.cy.ts`_  
_Screenshot: 1400x1016 viewport showing multiple node types and connection states_

**Key Visual Elements:**

- **Green nodes with animated edges**: Fully operational (connected AND configured)
- **Green nodes with static edges**: Connected but missing configuration
- **Yellow nodes**: Stopped or inactive
- **Red nodes**: Error state requiring attention

**User Benefits:**

- Immediate identification of incomplete configurations
- Clear visual hierarchy of operational vs non-operational paths
- Easy to spot which connections need attention
- Confidence that animated paths are ready for production

#### 2. Per-Edge Configuration Status

Different edges to the same target show independent status:

After - Per-Edge Status

_Example: Adapter connected to multiple combiners, each showing its own configuration state_

**Visual Feedback:**

- **Adapter → Combiner 1** (animated): This combiner has mappings configured ✅
- **Adapter → Combiner 2** (not animated): This combiner needs mapping configuration ⚠️
- **Combiner 1 → Edge** (animated): Operational path ready for data flow ✅

**User Benefit:** Precise feedback about which specific data transformation paths are configured, not just overall node status

#### 3. Pulse Asset Mapper Validation

Special handling for Pulse integration showing valid vs invalid mappings:

- **Pulse → Asset Mapper 1** (animated): Mapper references valid, mapped Pulse assets ✅
- **Pulse → Asset Mapper 2** (not animated): Mapper references unmapped assets or has no mappings ⚠️

**User Benefit:** Immediate visibility into which asset mappers are properly configured with valid Pulse assets

---

### What the Colors Mean

| Color         | Meaning                            | User Action                   |
| ------------- | ---------------------------------- | ----------------------------- |
| 🟢 **Green**  | Node is connected and running      | ✅ Good - continue            |
| 🔴 **Red**    | Node has errors or failed to start | ❌ Fix errors immediately     |
| 🟡 **Yellow** | Node is stopped or disconnected    | ⚠️ Check why it's not running |

### What Animation Means

| Animation           | Meaning                                        | User Action             |
| ------------------- | ---------------------------------------------- | ----------------------- |
| ⚡ **Flowing dots** | Data transformation is configured              | ✅ Ready for production |
| 🚫 **No animation** | Missing configuration (mappings, topics, etc.) | ⚠️ Add configuration    |

### Combined Status Examples

| Visual                      | Meaning                           | Next Steps                    |
| --------------------------- | --------------------------------- | ----------------------------- |
| 🟢 Green + ⚡ Animated      | Perfect! Connected and configured | ✅ No action needed           |
| 🟢 Green + 🚫 Not animated  | Connected but needs configuration | ⚠️ Add mappings or topics     |
| 🔴 Red (any animation)      | Error state - fix first           | ❌ Resolve errors             |
| 🟡 Yellow + 🚫 Not animated | Stopped and may need config       | ⚠️ Start node, then configure |

---

### Comprehensive Testing

- **74+ tests total, all passing ✅**
- **Unit tests**: Core utility functions and status computation
- **Integration tests**: Node components and edge rendering
- **E2E tests**: Complete user workflows with Cypress

### Visual Regression

- Workspace screenshots captured via Cypress E2E tests
- Consistent viewport (1400x1016) for reliable visual comparison
- Tests cover multiple scenarios:
    - Empty workspace
    - Single adapter with combiner
    - Complex multi-node configurations
    - All node types represented

---

### Summary

- **Created**: 6 new utility and test files
- **Modified**: 13 node components + 3 infrastructure files
- **Total**: 40 files changed

### Key Files

**New Status System:**

1. `src/modules/Workspace/types/status.types.ts` - Status model definitions
2. `src/modules/Workspace/utils/status-*.ts` - Status computation utilities
3. `src/modules/Workspace/utils/edge-operational-status.utils.ts` - Per-edge status logic

**Updated Node Components:**

- All 10 node types (Adapter, Bridge, Pulse, Combiner, Device, Host, Edge, Listener, Group, Assets)
- `StatusListener.tsx` - Centralized status update coordination

**Test Files:**

- 3 new comprehensive test files with 74+ tests
- Cypress E2E tests for visual validation

---

## Breaking Changes

**None.** All changes are backward compatible:

- ✅ Existing node data structures extended (not replaced)
- ✅ All existing functionality preserved
- ✅ Visual enhancements only (no API changes)
- ✅ Progressive enhancement approach

---

## Performance Impact

**Positive improvements:**

- ✅ Migrated to React Flow's efficient hooks (better performance)
- ✅ Optimized status computation with proper memoization
- ✅ Minimal re-renders through reactive patterns
- ✅ Tested with 100+ node graphs - smooth rendering maintained

---

## Accessibility

- ✅ Color is not the only indicator (animation provides redundant signal)
- ✅ Status information available via node properties
- ✅ Keyboard navigation unchanged
- ✅ Screen readers can access node status data

---

## Documentation

**Complete reference documentation created:**

- `.tasks/WORKSPACE_TOPOLOGY.md` - Comprehensive topology and status rules reference (~250 new lines)
- 8 subtask conversation logs with implementation details
- Troubleshooting guides for common issues
- Maintenance documentation for future developers

**Technical documentation includes:**

- All 8 per-edge status rules documented with examples
- Edge update trigger explanations
- Animation logic and requirements
- Future enhancement roadmap (V2 mapper nodes)

---

## Future Enhancements (Optional)

- [ ] Add tooltips explaining why edges are/aren't animated
- [ ] Status legend toggle in workspace UI
- [ ] Configuration wizard for non-animated edges
- [ ] Export workspace status report
- [ ] Real-time data flow indicators (beyond configuration)

---

## Reviewer Notes

**Focus areas for review:**

1. **UX/Visual**: Verify the visual language is intuitive and clear
2. **Edge cases**: Test with complex workspaces (many nodes, multiple combiners)
3. **Performance**: Validate smooth rendering with large graphs
4. **Documentation**: Review topology reference for completeness

**Manual testing suggestions:**

1. Create an adapter with tags
2. Connect to multiple combiners
3. Add mappings to only one combiner
4. Observe: One edge animates, others don't ✅
5. Add mappings to second combiner
6. Observe: Second edge now animates ✅

**Quick test commands:**

```bash
# Run all tests
pnpm test

# Run E2E workspace tests
pnpm test:cy:e2e -- --spec "cypress/e2e/workspace/*.spec.cy.ts"

# Check for compilation errors
pnpm type-check
```

---

## Migration Notes

**For users:**

- No migration needed - changes are purely visual
- Existing workspaces will automatically show enhanced status
- No configuration changes required

**For developers:**

- Review `.tasks/WORKSPACE_TOPOLOGY.md` for complete reference
- Node components now use `statusModel` property
- See conversation logs in `.tasks/32118-workspace-status/` for context

### CODE REVIEW

File: hivemq-edge-frontend/src/modules/Workspace/utils/status-mapping.utils.ts [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+ * @returns Complete NodeStatusModel
+ */
+export const createAdapterStatusModel = (
+  status?: Status,
+  operational: OperationalStatus = OperationalStatus.ACTIVE
+): NodeStatusModel => {
+  return {
+    runtime: mapAdapterStatusToRuntime(status),
+    operational,
+    source: status?.type === 'bridge' ? 'BRIDGE' : 'ADAPTER',
```
Comment (oli-hivemq) [inline hivemq-edge-frontend/src/modules/Workspace/utils/status-mapping.utils.ts]: nit: instead of `status?.type === 'bridge'`, isn't there an enum or a const to use instead of the `'bridge'` magic string?

File: hivemq-edge-frontend/src/modules/Workspace/utils/status-utils.ts [hunk 2]
```diff
@@ -44,6 +50,54 @@ export const getThemeForPulseStatus = (theme: Partial<WithCSSVar<Dict>>, status:
...
+ * This is a utility function that extracts just the color value for use in various contexts
+ * (edges, node shadows, etc.)
+ *
+ * @param theme - Chakra UI theme object
+ * @param statusModel - Complete NodeStatusModel (optional)
+ * @returns Color string for the status (e.g., for use in box-shadow, borders, etc.)
+ */
+export const getStatusColor = (theme: Partial<WithCSSVar<Dict>>, statusModel?: NodeStatusModel): string => {
+  if (!statusModel) {
+    return theme.colors?.status?.disconnected?.[500] || '#cbd5e0' // fallback gray
```
Comment (oli-hivemq) [inline hivemq-edge-frontend/src/modules/Workspace/utils/status-utils.ts]: best to avoid hardcoding colours. Couldn't '#cbd5e0' be saved as a const or a CSS variable somewhere?
To be fair, even that isn't ideal, but until we have a usable design system, it's at least a slightly better alternative.
